EinOpsGptLanguageModel.py

```diff
@@ -57,12 +60,15 @@ def forward(self, idx: torch.Tensor, targets: Optional[torch.Tensor] = None, blo
             T, device=self.device) + history_length]
 
         x = tok_emb + pos_emb  # [B, T, C]
-        for layer, (projections, layer_w_in, layer_w_out, mha_proj, layer_scale) in enumerate(zip(self.attention_kvq, self.w_in, self.w_out, self.out_proj, self.scale)):
+        for layer, (layer_kv_proj, layer_q_proj, layer_w_in, layer_w_out, layer_out_proj, layer_scale) in enumerate(zip(self.kv_proj, self.q_proj, self.w_in, self.w_out, self.out_proj, self.scale)):
+
             x = F.layer_norm(x, (C,), weight=layer_scale[0])
             kv_cache = blocks_kvcache[layer] if blocks_kvcache else None
 
-            # [B, T, C] @ [3, C, h, d] -> [3, B, h, T, d], S = 3
-            k, v, q = torch.einsum('btc,schd->sbhtd', x, projections)
+            q = einsum(x, layer_q_proj, 'b t c, c g num_kv d -> b g num_kv t d') # [B, g, num_kv_heads, T, d]
+
+            # [B, T, C] @ [2, C, num_kv_heads, d] -> [2, B, T, num_kv_heads, d]
+            k, v = einsum(x, layer_kv_proj, 'b t c, s c num_kv d -> s b num_kv t d') # 2 [B, num_kv_heads, T, d]
```

Comment:
change `num_kv` to `kv_head` as its more symbolic of what iteration of the sum you are at